### PR TITLE
Default Prometheus rules

### PR DESCRIPTION
**Pull request by ahma**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| API breaks?     | no
| Deprecations?   | no
| License         | Apache 2.0

### Checklist

- [x] Code meets the [Developer Guide](https://github.com/banzaicloud/developer-guide)
- [x] User guide and development docs updated (if needed)
- [x] Related Helm chart(s) updated (if needed)

### CODE REVIEW

File: pkg/resources/fluentd/prometheusrules.go [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+								Type:   intstr.String,
+								StrVal: `rate(fluentd_status_buffer_queue_length[5m]) > 0.5`,
+							},
+							For: "1m",
+							Labels: map[string]string{
+								"service":  "fluentd",
+								"severity": "critical",
+							},
+							Annotations: map[string]string{
+								"summary":     `fluentd node are critical`,
```
Comment (pepov) [inline pkg/resources/fluentd/prometheusrules.go:96]: summary should be different

File: pkg/resources/fluentd/prometheusrules.go [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+							Annotations: map[string]string{
+								"summary":     `fluentd node are failing`,
+								"description": `In the last 5 minutes, fluentd queues increased 30%. Current value is {{ "{{ $value }}" }}`,
+							},
+						},
+						{
+							Alert: "FluentdQueueLength",
+							Expr: intstr.IntOrString{
+								Type:   intstr.String,
+								StrVal: `rate(fluentd_status_buffer_queue_length[5m]) > 0.5`,
```
Comment (pepov) [inline pkg/resources/fluentd/prometheusrules.go:88]: we need a different query here, this is a gauge

File: pkg/resources/fluentd/prometheusrules.go [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+							Annotations: map[string]string{
+								"summary":     `fluentd cannot be scraped`,
+								"description": `Prometheus could not scrape {{ "{{ $labels.job }}" }} for more than 30 minutes`,
+							},
+						},
+						{
+							Alert: "FluentdQueueLength",
+							Expr: intstr.IntOrString{
+								Type:   intstr.String,
+								StrVal: `rate(fluentd_status_buffer_queue_length[5m]) > 0.3`,
```
Comment (pepov) [inline pkg/resources/fluentd/prometheusrules.go:72]: we need a different query here, this is a gauge

File: pkg/resources/fluentd/prometheusrules.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+							},
+							For: "1m",
+							Labels: map[string]string{
+								"service":  "fluentd",
+								"severity": "warning",
+							},
+							Annotations: map[string]string{
+								"summary":     `fluentd node are failing`,
```
Comment (pepov) [inline pkg/resources/fluentd/prometheusrules.go:80]: invalid summary

**Review by pepov — CHANGES_REQUESTED** [review]
job and namespace labels should be added to all queries to avoid accidental conflicts with other fluentd deployments

**Review by tarokkk — APPROVED** [review]

File: pkg/resources/fluentbit/prometheusrules.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+	"github.com/banzaicloud/operator-tools/pkg/reconciler"
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (r *Reconciler) prometheusRules() (runtime.Object, reconciler.DesiredState, error) {
+	if r.Logging.Spec.FluentbitSpec.Metrics != nil && r.Logging.Spec.FluentbitSpec.Metrics.PrometheusRules {
+		objectMetadata := r.FluentbitObjectMeta(fluentbitServiceName)
+		nsJobLabel := fmt.Sprintf("job=\"%s\", namespace=\"%s\"", objectMetadata.Name, objectMetadata.Namespace)
```
Comment (siliconbrain) [inline pkg/resources/fluentbit/prometheusrules.go:29]:
```suggestion
		nsJobLabel := fmt.Sprintf(`job="%s", namespace="%s"`, objectMetadata.Name, objectMetadata.Namespace)
```

File: pkg/resources/fluentd/prometheusrules.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	"github.com/banzaicloud/operator-tools/pkg/reconciler"
+	v1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+func (r *Reconciler) prometheusRules() (runtime.Object, reconciler.DesiredState, error) {
+	if r.Logging.Spec.FluentdSpec.Metrics != nil && r.Logging.Spec.FluentdSpec.Metrics.PrometheusRules {
+		objectMetadata := r.FluentdObjectMeta(ServiceName+"-metrics", ComponentFluentd)
+		nsJobLabel := fmt.Sprintf("job=\"%s\", namespace=\"%s\"", objectMetadata.Name, objectMetadata.Namespace)
```
Comment (siliconbrain) [inline pkg/resources/fluentd/prometheusrules.go:29]:
```suggestion
		nsJobLabel := fmt.Sprintf(`job="%s", namespace="%s"`, objectMetadata.Name, objectMetadata.Namespace)
```

File: pkg/resources/fluentbit/prometheusrules.go [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+			Spec: v1.PrometheusRuleSpec{
+				Groups: []v1.RuleGroup{{
+					Name: "fluentbit",
+					Rules: []v1.Rule{
+						{
+							Alert: "FluentbitTooManyErrors",
+							Expr: intstr.IntOrString{
+								Type:   intstr.String,
+								StrVal: fmt.Sprintf("rate(fluentbit_output_retries_failed_total{%s}[10m]) > 0", nsJobLabel),
+							},
```
Comment (siliconbrain) [inline pkg/resources/fluentbit/prometheusrules.go:42]:
```suggestion
							Expr: intstr.FromString(fmt.Sprintf("rate(fluentbit_output_retries_failed_total{%s}[10m]) > 0", nsJobLabel)),
```

File: pkg/resources/fluentbit/prometheusrules.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+					},
+				},
+				},
+			},
+		}, reconciler.StatePresent, nil
+	}
+	return &v1.PrometheusRule{
+		ObjectMeta: r.FluentbitObjectMeta(fluentbitServiceName),
+		Spec:       v1.PrometheusRuleSpec{},
+	}, reconciler.StateAbsent, nil
```
Comment (siliconbrain) [inline pkg/resources/fluentbit/prometheusrules.go:62]:
```suggestion
func (r *Reconciler) prometheusRules() (obj runtime.Object, state reconciler.DesiredState, err error) {
	obj = &v1.PrometheusRule{
		ObjectMeta: r.FluentbitObjectMeta(fluentbitServiceName),
	}
	state = reconciler.StateAbsent
	
	if r.Logging.Spec.FluentbitSpec.Metrics != nil && r.Logging.Spec.FluentbitSpec.Metrics.PrometheusRules {
		nsJobLabel := fmt.Sprintf("job=\"%s\", namespace=\"%s\"", obj.Name, obj.Namespace)

		obj.Spec.Groups = []v1.RuleGroup{
			{
				Name: "fluentbit",
				Rules: []v1.Rule{
					{
						Alert: "FluentbitTooManyErrors",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("rate(fluentbit_output_retries_failed_total{%s}[10m]) > 0", nsJobLabel),
						},
						For: "10m",
						Labels: map[string]string{
							"service":  "fluentbit",
							"severity": "warning",
						},
						Annotations: map[string]string{
							"summary":     `Fluentbit too many errors.`,
							"description": `Fluentbit ({{ $labels.instance }}) is erroring.`,
						},
					},
				},
			},
		}
		state = reconciler.StatePresent
	}
	return
```

File: pkg/resources/fluentd/prometheusrules.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+					},
+				},
+				},
+			},
+		}, reconciler.StatePresent, nil
+	}
+	return &v1.PrometheusRule{
+		ObjectMeta: r.FluentdObjectMeta(ServiceName, ComponentFluentd),
+		Spec:       v1.PrometheusRuleSpec{},
+	}, reconciler.StateAbsent, nil
```
Comment (siliconbrain) [inline pkg/resources/fluentd/prometheusrules.go:174]:
```suggestion
func (r *Reconciler) prometheusRules() (obj runtime.Object, state reconciler.DesiredState, err error) {
	obj = &v1.PrometheusRule{
		ObjectMeta: r.FluentdObjectMeta(ServiceName+"-metrics", ComponentFluentd),
	}
	state = reconciler.StateAbsent
	
	if r.Logging.Spec.FluentdSpec.Metrics != nil && r.Logging.Spec.FluentdSpec.Metrics.PrometheusRules {
		nsJobLabel := fmt.Sprintf("job=\"%s\", namespace=\"%s\"", obj.Name, obj.Namespace)

		obj.Spec.Groups = []v1.RuleGroup{
			{
				Name: "fluentd",
				Rules: []v1.Rule{
					{
						Alert: "FluentdNodeDown",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("up{%s} == 0", nsJobLabel),
						},
						For: "10m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "critical",
						},
						Annotations: map[string]string{
							"summary":     `fluentd cannot be scraped`,
							"description": `Prometheus could not scrape {{ "{{ $labels.job }}" }} for more than 30 minutes`,
						},
					},
					{
						Alert: "FluentdQueueLength",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("rate(fluentd_status_buffer_queue_length{%s}[5m]) > 0.3", nsJobLabel),
						},
						For: "1m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "warning",
						},
						Annotations: map[string]string{
							"summary":     `fluentd node are failing`,
							"description": `In the last 5 minutes, fluentd queues increased 30%. Current value is {{ "{{ $value }}" }}`,
						},
					},
					{
						Alert: "FluentdQueueLength",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("rate(fluentd_status_buffer_queue_length{%s}[5m]) > 0.5", nsJobLabel),
						},
						For: "1m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "critical",
						},
						Annotations: map[string]string{
							"summary":     `fluentd nodes buffer queue length are critical`,
							"description": `In the last 5 minutes, fluentd queues increased 50%. Current value is {{ "{{ $value }}" }}`,
						},
					},
					{
						Alert: "FluentdRecordsCountsHigh",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("sum(rate(fluentd_output_status_emit_records{%s}[5m])) BY (pod) >  (3 * sum(rate(fluentd_output_status_emit_records{%s}[15m])) BY (pod))", nsJobLabel, nsJobLabel),
						},
						For: "1m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "critical",
						},
						Annotations: map[string]string{
							"summary":     `fluentd records count are critical`,
							"description": `In the last 5m, records counts increased 3 times, comparing to the latest 15 min.`,
						},
					},
					{
						Alert: "FluentdRetry",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("increase(fluentd_status_retry_count{%s}[10m]) > 0", nsJobLabel),
						},
						For: "20m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "warning",
						},
						Annotations: map[string]string{
							"summary":     `Fluentd retry count has been  {{ "{{ $value }}" }} for the last 10 minutes`,
							"description": `Fluentd retry count has been  {{ "{{ $value }}" }} for the last 10 minutes`,
						},
					},
					{
						Alert: "FluentdOutputError",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("increase(fluentd_output_status_num_errors{%s}[10m]) > 0", nsJobLabel),
						},
						For: "1s",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "warning",
						},
						Annotations: map[string]string{
							"summary":     `There have been Fluentd output error(s) for the last 10 minutes`,
							"description": `Fluentd output error count is {{ "{{ $value }}" }} for the last 10 minutes`,
						},
					},
					{
						Alert: "FluentdBufferSize",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("node_filesystem_avail_bytes{mountpoint=\"/buffers\", %s} / node_filesystem_size_bytes{mountpoint=\"/buffers\", %s} * 100 < 10", nsJobLabel, nsJobLabel),
						},
						For: "10m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "warning",
						},
						Annotations: map[string]string{
							"summary":     `Fluentd buffer free capacity less than 10%.`,
							"description": `Fluentd buffer size capacity is {{ "{{ $value }}" }}% `,
						},
					},
					{
						Alert: "FluentdBufferSize",
						Expr: intstr.IntOrString{
							Type:   intstr.String,
							StrVal: fmt.Sprintf("node_filesystem_avail_bytes{mountpoint=\"/buffers\", %s} / node_filesystem_size_bytes{mountpoint=\"/buffers\" ,%s} * 100 < 5", nsJobLabel, nsJobLabel),
						},
						For: "10m",
						Labels: map[string]string{
							"service":  "fluentd",
							"severity": "critical",
						},
						Annotations: map[string]string{
							"summary":     `Fluentd buffer free capacity less than 5%`,
							"description": `Fluentd buffer size capacity is {{ "{{ $value }}" }}% `,
						},
					},
				},
			},
		}
		state = reconciler.StatePresent
	}
	return
```
I think maybe there was a typo here cause the two calls to `FluentdObjectMeta` had different arguments.

**Review by siliconbrain — APPROVED** [review]
Just suggestions, but please check the two calls to `FluentdObjectMeta` because I think they unintentionally have different arguments.

**Review by siliconbrain — APPROVED** [review]

**Pull request merged**